Initial draft

ccthomas opened the pull request.

Commit:
First draft, work in progress

Commit:
Practice Adding Commit

Commit:
Progress made on July 1st

jeanconn commented on ConvertDraft.py:
```diff
@@ -0,0 +1,108 @@
+    synctest2.append(ir[sfduhdrl+2])
+    synctest2.append(ir[sfduhdrl+3])
+    notexplained=1
+    if total( total(synctest1,SFDUsync,'-') , total(synctest2,VCDUsync,'-') , '+') == [0,0,0,0]:
```
Your approach here of making this almost identical in syntax and function to the IDL is neat!  You probably don't however, need to carry it quite that far.  For the port, I suspect you can do something like

```
if (synctest1 == SFDUsync) and (synctest2 == VCDUsync):
```

jeanconn commented on ConvertDraft.py:
```diff
@@ -0,0 +1,108 @@
+                idx[tdsn]=indexo #TODO: Fix Error, tdsn is out od range for idx index
+                count[tdsn] = ir[sfduhdrl+6]*65536L+ir[sfduhdrl+7]*256L+ir[sfduhdrl+8]
+
+                ert[tdsn] = (ir[42]*256L + ir[43] - 14610)* 84600**0+ (((ir[44]*256 + ir[45])*256 + ir[46])*256 + ir[47])/1**3+ (ir[48]*256 + ir[49])/1**6
```
This might be a case where you'd want to pass bytes 42 - 49 or whatever to a tiny subroutine to perform what look like some bit operations on them (I confess, haven't looked at this too closely yet) to get the final answer.

ccthomas replied:
I just looked over that code and realized something. In the IDL code it has
84600d0 and without thinking I converted that to 84600**0 because we
thought 'd' is an exponent. But that does not make sense to me, why would
anyone have a number, "84600", to the zero power?

On Fri, Jul 1, 2016 at 3:58 PM, Jean Connelly notifications@github.com
wrote:

> In ConvertDraft.py
> https://github.com/sot/pydarkcal/pull/2#discussion_r69348864:
> 
> > -        ch=[]
> > -        ch.append(ir[sfduhdrl+4])
> > -        ch.append(ir[sfduhdrl+5])
> > -        if ch[1] - DSNch[1] == 0:
> > -            testr = []
> > -            for i in range(10,1028):
> > -                testr.append( ir[sfduhdrl+i])
> > -            testn=[]
> > -            for n in testr:
> > -                if n!=170:
> > -                    testn.append(n)
> > -            if testn > 3:
> > -                idx[tdsn]=indexo #TODO: Fix Error, tdsn is out od range for idx index
> > -                count[tdsn] = ir[sfduhdrl+6]_65536L+ir[sfduhdrl+7]_256L+ir[sfduhdrl+8]
> >   +
> > -                ert[tdsn] = (ir[42]_256L + ir[43] - 14610)_ 84600**0+ (((ir[44]_256 + ir[45])_256 + ir[46])*256 + ir[47])/1**3+ (ir[48]_256 + ir[49])/1_*6
> 
> This might be a case where you'd want to pass bytes 42 - 49 or whatever to
> a tiny subroutine to perform what look like some bit operations on them (I
> confess, haven't looked at this too closely yet) to get the final answer.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/sot/pydarkcal/pull/2/files/2ac5d8c63e9f1a04f627f104a036d8245666208c#r69348864,
> or mute the thread
> https://github.com/notifications/unsubscribe/ALe4z5Od8tJdDs7LvqYMAGtMYWhe1xd8ks5qRXF5gaJpZM4JDQTX
> .

jeanconn replied:
First, I think 'd' is an indicator of exponential notation, not the exponent operator.  This would make it equivalent to "e" in some systems (2e5 = 2 \* 10 *\* 5).  So, if we interpret it that way in the IDL, you get 84600 \* 10 *\* 0, or basically 84600 back again.  The IDL trick here, however, seems to be that by specifying it in exponential notation, you get it back as a double:

```
IDL> print,typename(84600d0)
DOUBLE
IDL> print,typename(84600)
LONG
IDL> print,typename(84600.0)
FLOAT
```

This may or may not be required in the Python version, but I suspect that this line will need to be checked with real data.  A bigger concern I have is that the original IDL is using "84600d0" when there are "86400d0" seconds in a day (and I'm strongly assuming that this number is supposed to represent that).  But, I'm guessing that this calculated time is then not used for much if it is wrong and we haven't noticed.  For the port, you should continue with the "wrong" value of 84600 and we can look at fixing it later.

jeanconn replied:
@taldcroft I'm wondering if we should re-look at any historical dark cals that cross day boundaries... (actually, never mind, if these are 84600-second days, I would not immediately be able to tell which dark cals would have pieces that would cross those boundaries.)

ccthomas replied:
I just checked all the (.pro) files and this number is not used again. It
is used only in that one line of code.

On Tue, Jul 5, 2016 at 11:57 AM, Jean Connelly notifications@github.com
wrote:

> In ConvertDraft.py
> https://github.com/sot/pydarkcal/pull/2#discussion_r69589344:
> 
> > -        ch=[]
> > -        ch.append(ir[sfduhdrl+4])
> > -        ch.append(ir[sfduhdrl+5])
> > -        if ch[1] - DSNch[1] == 0:
> > -            testr = []
> > -            for i in range(10,1028):
> > -                testr.append( ir[sfduhdrl+i])
> > -            testn=[]
> > -            for n in testr:
> > -                if n!=170:
> > -                    testn.append(n)
> > -            if testn > 3:
> > -                idx[tdsn]=indexo #TODO: Fix Error, tdsn is out od range for idx index
> > -                count[tdsn] = ir[sfduhdrl+6]_65536L+ir[sfduhdrl+7]_256L+ir[sfduhdrl+8]
> >   +
> > -                ert[tdsn] = (ir[42]_256L + ir[43] - 14610)_ 84600**0+ (((ir[44]_256 + ir[45])_256 + ir[46])*256 + ir[47])/1**3+ (ir[48]_256 + ir[49])/1_*6
> 
> @taldcroft https://github.com/taldcroft I'm wondering if we should
> re-look at any historical dark cals that cross day boundaries...
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/sot/pydarkcal/pull/2/files/2ac5d8c63e9f1a04f627f104a036d8245666208c#r69589344,
> or mute the thread
> https://github.com/notifications/unsubscribe/ALe4z9S84AWB6Hw4P3njaerq6DugPfQnks5qSn7wgaJpZM4JDQTX
> .

jeanconn replied:
By "number" do you mean 84600 or the value that is calculated and saved as ert[tdsn]?  Because I assume the ert value is used all over the place.

ccthomas replied:
Yes I meant "84600", ert is easily used 50-80 times across all of the files

On Tue, Jul 5, 2016 at 12:06 PM, Jean Connelly notifications@github.com
wrote:

> In ConvertDraft.py
> https://github.com/sot/pydarkcal/pull/2#discussion_r69591009:
> 
> > -        ch=[]
> > -        ch.append(ir[sfduhdrl+4])
> > -        ch.append(ir[sfduhdrl+5])
> > -        if ch[1] - DSNch[1] == 0:
> > -            testr = []
> > -            for i in range(10,1028):
> > -                testr.append( ir[sfduhdrl+i])
> > -            testn=[]
> > -            for n in testr:
> > -                if n!=170:
> > -                    testn.append(n)
> > -            if testn > 3:
> > -                idx[tdsn]=indexo #TODO: Fix Error, tdsn is out od range for idx index
> > -                count[tdsn] = ir[sfduhdrl+6]_65536L+ir[sfduhdrl+7]_256L+ir[sfduhdrl+8]
> >   +
> > -                ert[tdsn] = (ir[42]_256L + ir[43] - 14610)_ 84600**0+ (((ir[44]_256 + ir[45])_256 + ir[46])*256 + ir[47])/1**3+ (ir[48]_256 + ir[49])/1_*6
> 
> By "number" do you mean 84600 or the value that is calculated and saved as
> ert[tdsn]? Because I assume the ert value is used all over the place.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/sot/pydarkcal/pull/2/files/2ac5d8c63e9f1a04f627f104a036d8245666208c#r69591009,
> or mute the thread
> https://github.com/notifications/unsubscribe/ALe4zy1GPLxgkf37rXTQCiy4tC1SUWVmks5qSoEMgaJpZM4JDQTX
> .

Commit:
Progress made on July 6th

Commit:
Progress made on July 6th

Commit:
Fixed Indexing

jeanconn commented on ConvertDraft.py:
```diff
@@ -0,0 +1,105 @@
+
+
+r = []
+for i in range(len(il)/sfdurecl):
```
Can you describe for me what this is doing?

ChristopherThomas94 replied:
>f = open(data_file,'r+b')
>il = f.read()
>f.close()
>ol = open('../data2/test_file_python.txt','w')
'r+b' opens a file in binary mode. I save that to il, which is the same variable name (Robert Cameron?) used. I then open a different file to be written to name ol. Essentially the same as the .anal file, but I gave it a different file name to save to for now


>prb(ol,['ACA cal data file:',data_file])
>prb(ol,'')
>prb(ol,['Valid SFDU sync is:',str(SFDUsync)])
>prb(ol,['Valid DSN record sync is:',str(VCDUsync),str(DSNch)])
>prb(ol,'')
prb: prb was a .pro file used to print both to the .anal file and to the terminal. I turned prb.pro into a  function found at the beginning of this file. prb.pro took anywhere from 2-28 arguments. To overcome this transition to python I gave the prb function two arguments. first argument is the same as prb.pro, which is the file to be written to. The last argument in prb.pro was a format option, I got rid of that for now because it seemed useless in python. The other 26 args, a-z, where combined to make a string to be printed. I decided to have the prb function take in an array, which holds a-z, and then combines the array within the function to be printed.

Arguments In This Example: The arguments I am passing in (data_file,SFDUsync,VCDUsync,DSNch), is the name of the file being read and the three others are variables defined in variable.py. In the .pro file they have already been initialized to the value shown in Convert.py earlier in the code. The three variables are not touched between initialization and being passed into prb to be printed.

jeanconn replied:
Hi Christopher, I actually meant to describe the content of this one line above

```
for i in range(len(il)/sfdurec1):
```

 (though your description of the opening and printing would be great as just long comments in the code, so this wasn't wasted effort).

taldcroft replied:
You should do this using the Python logging system, which allows highly configurable logging of output to multiple destinations.  The native logging module is a little annoying to work with, so we have a wrapper that does just what you want:  http://cxc.harvard.edu/contrib/pyyaks/logger.html

For example:
```
import pyyaks.logger
logger = pyyaks.logger.get_logger(logfilename)
...
logger.info('Valid DSN record sync is: %s %s', VCDUsync, DSNch)
```

Commit:
Code converted up to
> Processing DSN record segment 1 <

jeanconn commented on ConvertDraft.py:
```diff
@@ -0,0 +1,105 @@
+print 'Target for SFDUs: ' + `i`
+
+#pro prb, olun,a,b,c,...,z,format=fmt
+def prb(olun,param_array):
```
(Moving Tom's excellent comment about logging as a comment on this prb function).

You should do this using the Python logging system, which allows highly configurable logging of output to multiple destinations.  The native logging module is a little annoying to work with, so we have a wrapper that does just what you want:  http://cxc.harvard.edu/contrib/pyyaks/logger.html

For example:

```
import pyyaks.logger
logger = pyyaks.logger.get_logger(logfilename)
...
logger.info('Valid DSN record sync is: %s %s', VCDUsync, DSNch)
```

jeanconn replied:
@taldcroft, @ChristopherThomas94 is still working in an Anaconda Python.  Should he pip install pyyaks for this?

taldcroft replied:
Yes, he should be able to `pip install pyyaks` now.  (I just uploaded it to PyPI).

jeanconn commented on ConvertDraft.py:
```diff
@@ -0,0 +1,105 @@
+prb(ol,['Valid SFDU sync is:',str(SFDUsync)])
+prb(ol,['Valid DSN record sync is:',str(VCDUsync),str(DSNch)])
+prb(ol,'')
+
```
(Moving Christopher's comment on the code up through line 34 as this new github comment)

Apparently I was looking at a preview of what you wanted me to explain (What I have below) and did not see all the code you highlighted. I will add comments and edit my code to fit the style guide and push all of it sometime tomorrow.

**f = open(data_file,'r+b')
il = f.read()
f.close()
ol = open('../data2/test_file_python.txt','w')**

'r+b' opens a file in binary mode. I save that to il, which is the same variable name (Robert Cameron?) used. I then open a different file to be written to name ol. Essentially the same as the .anal file, but I gave it a different file name to save to for now

**prb(ol,['ACA cal data file:',data_file])
prb(ol,'')
prb(ol,['Valid SFDU sync is:',str(SFDUsync)])
prb(ol,['Valid DSN record sync is:',str(VCDUsync),str(DSNch)])
prb(ol,'')**

prb: prb was a .pro file used to print both to the .anal file and to the terminal. I turned prb.pro into a  function found at the beginning of this file. prb.pro took anywhere from 2-28 arguments. To overcome this transition to python I gave the prb function two arguments. first argument is the same as prb.pro, which is the file to be written to. The last argument in prb.pro was a format option, I got rid of that for now because it seemed useless in python. The other 26 args, a-z, where combined to make a string to be printed. I decided to have the prb function take in an array, which holds a-z, and then combines the array within the function to be printed.

Arguments In This Example: The arguments I am passing in (data_file,SFDUsync,VCDUsync,DSNch), is the name of the file being read and the three others are variables defined in variable.py. In the .pro file they have already been initialized to the value shown in Convert.py earlier in the code. The three variables are not touched between initialization and being passed into prb to be printed.

Commit:
styled code to PEP8, Also added Comments
note: I do not know what a lot of the code does, I just know the outputs for both idl and python match. So for those comments I just copied and pasted what was on the idl file to the pyhton file

jeanconn commented on ConvertDraft.py:
```diff
@@ -0,0 +1,155 @@
+il = open(data_file, 'r+b').read()
+
+r = []
+for i in range(len(il)/sfdurecl):
```
The way I read this code, you are breaking the file up into N 1134 byte chunks and then checking each chunk to see if it matches the expected pattern.  So, you're checking that the chunk starting at byte 1 matches and then you're checking that the chunk starting at byte 1135 matches etc, right?  I think the original code checks the 1134 byte chunk starting at the file's byte 1, then the 1134 byte chunk that starts at byte 2 of the file, then the chunk that starts at byte 3, etc.

jeanconn replied:
Updating this comment for the record.  It looks like ASSOC does work the way you read it and dividing the file up in to N chunks where N is the length of the file divided by the length of an sfdu does seem fine.  The strategy of making a test array "r" does create one additional copy of the file in memory, but that should still be acceptable at these files sizes.

Commit:
Fixed some style guide errors

Commit:
variable disc problems, Jul28

jeanconn commented on ConvertDraft.py:
```diff
@@ -123,39 +215,138 @@ def prb(olun, param_array):
+# ==========================================================================
+# ==========================================================================
+# disc equals count - shifted_count at all elements not equals to 1
 disc = [i for i in (count - np.roll(count, 1)) if i != 1]
```
The np.roll is great to do the shift operation.  I don't immediately understand _why_ the IDL is doing that, but that's another story.   You don't need to loop over the array to get the values though, you can do this more as a numpy array operation with something like:

```
disc = np.flatnonzero(count - np.roll(count, 1) != 1)
```

I think that flatnonzero in this case will plug in OK for IDL's `where`.

jeanconn replied:
PS I think your original loop was giving you values != 1 when the IDL `where` was probably returning indexes.

ccthomas replied:
You are right about the indexes. And using
"disc = np.flatnonzero(count - np.roll(count, 1) != 1)"
fixed that problem. disc are now equal in IDL and Python

jeanconn replied:
Great!

jeanconn commented on ConvertDraft.py:
```diff
@@ -1,5 +1,89 @@
+# idx=lonarr(nrecmax)
+idx = []
+# count=lonarr(nrecmax)
+count = []
```
If you want to keep the same logic to make `disc` (with the shift/roll), you might need to start by making this this list into a pre-initialized array with all of the same zeros as the one in the IDL.

jeanconn replied:
Actually, nevermind, it looks like the logic later that does
```
count=count(0:tdsn-1)
``` 
would trim all the zeros off the end anyway.

Commit:
Fixed disc

Commit:
Problems with maindat 2Darray

Commit:
fixed maindat

Commit:
updating ConvertDraft & adding acalimg.pro

ccthomas committed:
Create README.md

jeanconn commented:
:+1:
